File: scripts/serve.js
#!/usr/bin/env node
/**
 * Simple HTTP server to serve the wallet approval HTML file.
 * This ensures proper HTTP context for wallet extensions.
 */

const http = require('http');
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');

const PORT = process.argv[2] || 8000;
const HOST = 'localhost';

// MIME types for different file extensions
const mimeTypes = {
    '.html': 'text/html',
    '.js': 'text/javascript',
    '.css': 'text/css',
    '.json': 'application/json',
    '.png': 'image/png',
    '.jpg': 'image/jpg',
    '.gif': 'image/gif',
    '.svg': 'image/svg+xml',
    '.wav': 'audio/wav',
    '.mp4': 'video/mp4',
    '.woff': 'application/font-woff',
    '.ttf': 'application/font-ttf',
    '.eot': 'application/vnd.ms-fontobject',
    '.otf': 'application/font-otf',
    '.wasm': 'application/wasm'
};

const server = http.createServer((req, res) => {
    // Add CORS headers for wallet compatibility
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
    
    if (req.method === 'OPTIONS') {
        res.writeHead(200);
        res.end();
        return;
    }
    
    let filePath = '.' + req.url;
    if (filePath === './') {
        filePath = './wallet-approval-enhanced-enhanced.html';
    }
    
    const extname = String(path.extname(filePath)).toLowerCase();
    const mimeType = mimeTypes[extname] || 'application/octet-stream';
    
    fs.readFile(filePath, (error, content) => {
        if (error) {
            if (error.code === 'ENOENT') {
                res.writeHead(404, { 'Content-Type': 'text/html' });
                res.end(`
                    <h1>404 - File Not Found</h1>
                    <p>The file <code>${req.url}</code> was not found.</p>
                    <p><a href="/wallet-approval.html">Go to Wallet Interface</a></p>
                `);
            } else {
                res.writeHead(500);
                res.end(`Server Error: ${error.code}`);
            }
        } else {
            res.writeHead(200, { 'Content-Type': mimeType });
            res.end(content, 'utf-8');
        }
    });
});

// Check if wallet-approval.html exists
if (!fs.existsSync('./wallet-approval-enhanced.html')) {
    console.log('❌ Error: wallet-approval.html not found in current directory');
    console.log(`   Current directory: ${process.cwd()}`);
    process.exit(1);
}

server.listen(PORT, HOST, () => {
    const url = `http://${HOST}:${PORT}/wallet-approval-enhanced.html`;
    
    console.log('🚀 Starting local development server...');
    console.log(`📁 Serving files from: ${process.cwd()}`);
    console.log(`🌐 Server running at: http://${HOST}:${PORT}`);
    console.log(`🔗 Wallet interface: ${url}`);
    console.log('');
    console.log('✅ This should fix MetaMask detection issues!');
    console.log('💡 Press Ctrl+C to stop the server');
    console.log('');
    
    // Try to open browser automatically
    const start = process.platform === 'darwin' ? 'open' : 
                  process.platform === 'win32' ? 'start' : 'xdg-open';
    
    exec(`${start} ${url}`, (error) => {
        if (error) {
            console.log('⚠️  Please manually open:', url);
        } else {
            console.log('🌐 Opening browser automatically...');
        }
    });
    
    console.log('=' .repeat(50));
});

server.on('error', (error) => {
    if (error.code === 'EADDRINUSE') {
        console.log(`❌ Port ${PORT} is already in use. Try a different port:`);
        console.log(`   node serve.js ${PORT + 1}`);
    } else {
        console.log(`❌ Error starting server: ${error.message}`);
    }
});

process.on('SIGINT', () => {
    console.log('\n\n👋 Server stopped. Goodbye!');
    process.exit(0);
}); 

File: scripts/serve.py
#!/usr/bin/env python3
"""
Simple HTTP server to serve the wallet approval HTML file.
This ensures proper HTTPS/HTTP context for wallet extensions.
"""

import http.server
import os
import socketserver
import sys
import webbrowser

# Configuration
PORT = 8000
DIRECTORY = "."

class Handler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=DIRECTORY, **kwargs)
    
    def end_headers(self):
        # Add CORS headers for wallet compatibility
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        super().end_headers()

def main():
    # Change to the script directory
    script_dir = os.path.dirname(os.path.abspath(__file__))
    os.chdir(script_dir)
    
    # Check if wallet-approval.html exists
    if not os.path.exists('wallet-approval.html'):
        print("❌ Error: wallet-approval.html not found in current directory")
        print(f"   Current directory: {os.getcwd()}")
        sys.exit(1)
    
    try:
        with socketserver.TCPServer(("", PORT), Handler) as httpd:
            url = f"http://localhost:{PORT}/wallet-approval.html"
            
            print("🚀 Starting local development server...")
            print(f"📁 Serving files from: {os.getcwd()}")
            print(f"🌐 Server running at: http://localhost:{PORT}")
            print(f"🔗 Wallet interface: {url}")
            print()
            print("✅ This should fix MetaMask detection issues!")
            print("💡 Press Ctrl+C to stop the server")
            print()
            
            # Try to open browser automatically
            try:
                webbrowser.open(url)
                print("🌐 Opening browser automatically...")
            except:
                print("⚠️  Please manually open:", url)
            
            print("\n" + "="*50)
            httpd.serve_forever()
            
    except KeyboardInterrupt:
        print("\n\n👋 Server stopped. Goodbye!")
    except OSError as e:
        if e.errno == 48:  # Address already in use
            print(f"❌ Port {PORT} is already in use. Try a different port:")
            print(f"   python3 serve.py {PORT + 1}")
        else:
            print(f"❌ Error starting server: {e}")

if __name__ == "__main__":
    # Allow custom port as command line argument
    if len(sys.argv) > 1:
        try:
            PORT = int(sys.argv[1])
        except ValueError:
            print("❌ Invalid port number. Using default port 8000.")
    
    main() 